posthog/hogql/functions/mapping.py

```diff
@@ -325,7 +325,9 @@ def compare_types(arg_types: list[ConstantType], sig_arg_types: tuple[ConstantTy
     "arrayDistinct": HogQLFunctionMeta("arrayDistinct", 1, 1),
     "arrayEnumerateDense": HogQLFunctionMeta("arrayEnumerateDense", 1, 1),
     "arrayIntersect": HogQLFunctionMeta("arrayIntersect", 1, None),
-    # "arrayReduce": HogQLFunctionMeta("arrayReduce", 2,None),  # takes a "parametric function" as first arg, is that safe?
```

Comment:
@Twixes looks like this comment was added by you. Do you know if it's safe to add the `arrayReduce` function yet?

Reply:
I _think_ yes, but it was actually @mariusandra in [here](https://github.com/PostHog/posthog/commit/692fa66e6390af9bb45c38213b913e4ef22f729e#diff-5f19b03cbc6b7f294cabe74afca9b139317c28f3534f35bf8c1bac351ef6ed26R58-R59), so I'll defer to him

Reply:
Oh dear... hmm... those might be safe, but we don't know. Somebody needs to dig into the ClickHouse C++ code and verify. I haven't had the time to do that yet.

My biggest worry is of course that users will be able to call things that they shouldn't, and such "magic function name constants" share the 💩 out of me: `arrayReduce('sumMap', ...)`. Other than `sumMap`, can we be 100% certain that anything users pass in there is safe? I can't.

If we enable this function, we'll at the very least need to be extremely prudent about what strings we allow as the first argument of the function, and ideally even make sure that the arguments passed match a certain format (min/max length?).

Reply:
I think the easiest solution is to enable `arrayReduce`, but limit the first argument to a list of allowed constants. It's not as hard to implement as you might think: all we need is [some logic in printer.py](https://github.com/PostHog/posthog/blob/ec59e6f43d6a5550548db1cf7ce2b1e018a3e854/posthog/hogql/printer.py#L716) that makes sure that for it, the first argument must equals `ast.Constant(value='sumMap')` or any of the other allowed constants...

Reply:
Prepped this in https://github.com/PostHog/posthog/pull/23261 (it really wasn't that difficult)

Reply:
Resolved in https://github.com/PostHog/posthog/commit/7d7784251adbd3077cfbb66826de2d6e6cd95f7f#r143578812